src/payment/payjoin/handler.rs

```diff
@@ -0,0 +1,433 @@
+
+use crate::config::PAYJOIN_REQUEST_TIMEOUT;
+use crate::error::Error;
+use crate::event::PayjoinPaymentFailureReason;
+use crate::logger::FilesystemLogger;
+use crate::payment::store::PaymentDetailsUpdate;
+use crate::payment::PaymentKind;
+use crate::payment::{PaymentDirection, PaymentStatus};
+use crate::types::{ChainSource, EventQueue, PaymentStore, Wallet};
+use crate::Event;
+use crate::PaymentDetails;
+
+use lightning::chain::channelmonitor::ANTI_REORG_DELAY;
+use lightning::chain::transaction::TransactionData;
+use lightning::chain::{Confirm, Filter, WatchedOutput};
+use lightning::ln::channelmanager::PaymentId;
+use lightning::log_error;
+use lightning::util::logger::Logger;
+
+use std::sync::{Arc, RwLock};
+
+pub(crate) struct PayjoinHandler {
+	chain_source: Arc<ChainSource>,
+	event_queue: Arc<EventQueue>,
+	logger: Arc<FilesystemLogger>,
+	payjoin_relay: payjoin::Url,
+	payment_store: Arc<PaymentStore>,
+	transactions: RwLock<Vec<PayjoinTransaction>>,
+	wallet: Arc<Wallet>,
+}
+
+impl PayjoinHandler {
+	pub(crate) fn new(
+		chain_source: Arc<ChainSource>, event_queue: Arc<EventQueue>,
+		logger: Arc<FilesystemLogger>, payjoin_relay: payjoin::Url,
+		payment_store: Arc<PaymentStore>, wallet: Arc<Wallet>,
+	) -> Self {
+		Self {
+			chain_source,
+			event_queue,
+			logger,
+			payjoin_relay,
+			payment_store,
+			transactions: RwLock::new(Vec::new()),
+			wallet,
+		}
+	}
+
+	pub(crate) fn start_request(
+		&self, payjoin_uri: payjoin::Uri<'_, NetworkChecked>,
+	) -> Result<Psbt, Error> {
+		let amount = payjoin_uri.amount.ok_or(Error::PayjoinRequestMissingAmount)?;
+		let receiver = payjoin_uri.address.clone();
+		let original_psbt =
+			self.wallet.build_payjoin_transaction(amount, receiver.clone().into())?;
+		self.transactions.write().unwrap().push(PayjoinTransaction::new(
+			original_psbt.clone(),
+			receiver.clone(),
+			amount,
+		));
+		let payment_id = original_psbt.unsigned_tx.txid()[..].try_into().map_err(|_| {
+			log_error!(self.logger, "Failed to start Payjoin request: invalid payment id");
+			Error::PayjoinRequestSendingFailed
+		})?;
+		self.payment_store.insert(PaymentDetails::new(
+			PaymentId(payment_id),
+			PaymentKind::Payjoin,
+			Some(amount.to_sat()),
+			PaymentDirection::Outbound,
+			PaymentStatus::Pending,
+		))?;
+		let tx = original_psbt.clone().unsigned_tx;
+		// watch original transaction in case it gets broadcasted by the receiver
+		self.register_tx(&tx.txid(), Script::empty());
+		self.event_queue.add_event(Event::PayjoinPaymentPending {
+			txid: tx.txid(),
+			amount: amount.to_sat(),
+			receipient: receiver.into(),
+		})?;
+		Ok(original_psbt)
+	}
+
+	pub(crate) async fn send_request(
+		&self, payjoin_uri: payjoin::Uri<'_, NetworkChecked>, original_psbt: &mut Psbt,
+	) -> Result<Option<Psbt>, Error> {
+		let (request, context) = payjoin::send::RequestBuilder::from_psbt_and_uri(
+			original_psbt.clone(),
+			payjoin_uri.clone(),
+		)
+		.and_then(|b| b.build_non_incentivizing())
```

Comment:
There are other options here, as `build_recommended` and others, I am looking into them to see whats the best to our needs(regular pj tx or channel opening)